### ISSUE TITLE
Static Extensions (née Resources)

### ISSUE THREAD

**Issue by ss2165** [CQCL/hugr#395]
It would be useful to have commonly used extension types and operations (e.g. from the Prelude) defined statically. 

Currently the `Resource` is defined with dynamically allocated data structures. Further it only stores `TypeDef`s and `OpDef`s which have to be dynamically allocated. 

I propose an alternative mechanism for statically declaring Extensions that hold only unparameterised (no TypeParam) types and ops, which can then also declare static references to those elements.

This may also make it easier to avoid string comparisons for type equality checking, but to what extent is unclear (not sure what Rust will let us get away with).

**Comment by ss2165** [comment]
I think the current mix of `lazy_static` + constant `Type::new_extension(CustomType::new_simple...` + constant string identifiers + `Arc<OpDef>` usage is good enough and the full thing set out here is over-optimisation. I would suggest lowering the priority of this, or adjusting the scope to be just adding unparameterised OpDef construction (akin to `Customtype::new_simple` etc.). @aborgna-q ?

**Comment by aborgna-q** [comment]
Agreed. I'd leave it for future perf optimizations; this isn't `P-high` anymore.
